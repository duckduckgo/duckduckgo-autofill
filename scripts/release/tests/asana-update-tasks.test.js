let mockAsanaClient;
jest.mock('asana', () => {
    const notes = `<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li>[[pr_url]]</li>
    <li>[[extra_content]]</li>
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>`;

    mockAsanaClient = {
        useAccessToken: () => mockAsanaClient,
        tasks: {
            getTask: () => ({ html_notes: notes }),
            updateTask: jest.fn(),
        },
    };
    return { Client: { create: () => mockAsanaClient } };
});

beforeEach(() => {
    process.env.ASANA_ACCESS_TOKEN = 'mock-asana-access-token';
    process.env.ANDROID_PR_URL = 'https://github.com/duckduckgo/android/pr/1';
    process.env.APPLE_PR_URL = 'https://github.com/duckduckgo/apple-browsers/pr/1';
    process.env.WINDOWS_PR_URL = 'https://github.com/duckduckgo/windows-browser/pr/1';
    process.env.ASANA_OUTPUT = JSON.stringify({
        android: {
            displayName: 'Android',
            taskGid: 'android-123',
            taskUrl: 'https://example.com/android-task',
        },
        apple: {
            displayName: 'Apple',
            taskGid: 'apple-123',
            taskUrl: 'https://example.com/apple-task',
        },
        extensions: {
            displayName: 'Extensions',
            taskGid: 'extension-123',
            taskUrl: 'https://example.com/extension-task',
        },
        windows: {
            displayName: 'Windows',
            taskGid: 'windows-123',
            taskUrl: 'https://example.com/windows-task',
        },
    });
});

describe('when Asana output is not available', () => {
    beforeEach(() => {
        process.env.ASANA_OUTPUT = undefined;
    });

    it('displays an error message', async () => {
        // @ts-ignore
        const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {});
        const mockConsole = jest.spyOn(console, 'error').mockImplementation(() => {});
        jest.isolateModules(() => {
            require('../asana-update-tasks.js');
        });
        // Wait for update tasks to run
        await new Promise((resolve) => setTimeout(resolve, 200));

        expect(mockExit).toHaveBeenCalledTimes(1);
        expect(mockConsole).toHaveBeenCalledTimes(1);
        expect(mockConsole.mock.calls).toMatchInlineSnapshot(`
      [
        [
          [Error: Unable to parse Asana output JSON],
        ],
      ]
    `);
    });
});

describe('when not all PR URLs are available', () => {
    beforeEach(() => {
        delete process.env.IOS_PR_URL;
    });

    it('updates each platforms task notes except for missing platforms', async () => {
        jest.isolateModules(() => {
            require('../asana-update-tasks.js');
        });
        // Wait for update tasks to run
        await new Promise((resolve) => setTimeout(resolve, 200));

        expect(mockAsanaClient.tasks.updateTask).toHaveBeenCalledTimes(3);
        expect(mockAsanaClient.tasks.updateTask.mock.calls).toMatchSnapshot();
    });
});

describe('when all data is available', () => {
    it('updates each platforms task notes', async () => {
        jest.isolateModules(() => {
            require('../asana-update-tasks.js');
        });
        // Wait for update tasks to run
        await new Promise((resolve) => setTimeout(resolve, 200));

        expect(mockAsanaClient.tasks.updateTask).toHaveBeenCalledTimes(3);
        expect(mockAsanaClient.tasks.updateTask.mock.calls).toMatchSnapshot();
    });
});
