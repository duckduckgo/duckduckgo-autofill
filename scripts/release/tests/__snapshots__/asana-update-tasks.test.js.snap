// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when all data is available updates each platforms task notes 1`] = `
[
  [
    "android-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/android/pr/1">Android PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "bsk-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/BrowserServicesKit/pr/1">BrowserServicesKit PR</a></li>
    <li><a href="https://github.com/duckduckgo/ios/pr/1">iOS PR</a></li><li><a href="https://github.com/duckduckgo/macos-browser/pr/1">macOS PR</a></li>
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "ios-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/ios/pr/1">iOS PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "macos-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/macos-browser/pr/1">macOS PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "windows-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/windows-browser/pr/1">Windows PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
]
`;

exports[`when not all PR URLs are available updates each platforms task notes except for missing platforms 1`] = `
[
  [
    "android-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/android/pr/1">Android PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "bsk-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/BrowserServicesKit/pr/1">BrowserServicesKit PR</a></li>
    <li>Error creating iOS PR</li><li><a href="https://github.com/duckduckgo/macos-browser/pr/1">macOS PR</a></li>
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "macos-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/macos-browser/pr/1">macOS PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
  [
    "windows-123",
    {
      "html_notes": "<h2>New release:<h2>
<ol>
    <li>version [[version]]</li>
    <li>[[commit]]</li>
    <li>[[release_url]]</li>
    <li><a href="https://github.com/duckduckgo/windows-browser/pr/1">Windows PR</a></li>
    
</ol>
<h2>Release notes:</h2>
<p>[[notes]]</p>",
    },
  ],
]
`;
